### ISSUE TITLE
Unrealistic temperatures when using restart file for  a run

### ISSUE THREAD

**Issue by hmkhatri** [NOAA-GFDL/MOM6#1157]
I am running an idealized channel model in ALE Z* mode (on a lat-lon grid) forced with steady winds, surface temperature restoration, and sponge at the north. The model runs fine if I start a new simulation. However, if I continue the simulation from a restart file (does not matter at what simulation time), temperatures increase drastically (see the attached domain-mean temperature and surface temperature plots, I continued the run from 7300 days using a restart file).  The data in the restart file looks fine. I do not see any missing or nan values. I have tested with the latest version of MOM6 as well as the early Jan version of MOM6 but the same issue appears. I have run multiple tests with different configurations. Switching off forcing terms, sponge did not help either.
![snapshot](https://user-images.githubusercontent.com/17830036/87069092-9c777480-c1e4-11ea-8133-5c6c7f9f5981.png)
![T](https://user-images.githubusercontent.com/17830036/87069093-9d100b00-c1e4-11ea-9d78-7f82b8374388.png)

Any idea what could be going wrong?

Issue #850 seems to report a similar problem. I have set VERTEX_SHEAR=True.  Could this be causing the problem?

**Comment by Hallberg-NOAA** [comment]
This problem seems to me to be unlikely to be due to VERTEX_SHEAR, but you could also check on this specific question by trying it out in a run with VERTEX_SHEAR=False and see whether the problem goes away. If it does not, we have well-established techniques for tracking down the source of non-reproducibility across restarts, which I will describe briefly.

  The first thing that you will want to do to debug your experiment is to find a cheap and quick case that reproduces these problems. Ideally you will have a case that can be run interactively within less than a minute or so.  Try seeing whether a 2-day run reproduces a pair of 1-day runs with a restart in the middle. If this case is too large to conveniently run interactively, try a lower resolution version of this case to see whether it exhibits the same pathology of differing across restarts.

 Once you have your cheap test case, you will want to set DEBUG=True in your MOM_override or MOM_input file and then repeat the two runs with and without a restart in the middle.  This generates a lot of debugging output to stdout, so you will want to direct this to a file.  For the pair of 1-day runs, append the output of the second run to the first, and then use a tool like `tkdiff` to compare this with the output of a single 2-day run.  You might find that these files are inconveniently large, in which case `head` and `tail` can be use to extract smaller subsets of these files that are supposed to correspond to one another.  The file from pair of runs will have some added stuff in the middle associated with the termination and restart of the runs, but it after this it is supposed to be identical with the output from the single longer run. You might find it convenient to edit this file to remove these extra lines before comparing the files. The first point where the chksums differ should have a description that gives you a good idea of where in the code or in which field the divergence arises.

  If you find your are having trouble with this, reach out to me and I can help you.

**Comment by hmkhatri** [comment]
@Hallberg-NOAA I followed your instructions. I ran one 2-day simulation and pair of 1-day runs with a restart in the middle. I compared the output in the debug mode and I could find the lines where the output starts to differ. The differences are first seen in forcing `fluxes%vprec` and `Beginning of step_MOM  h` lines. The output starts to diverge as soon as the run starts from the restart file. In the output below, the first line is the output from the original 2-day run and the second line is from the run with the restart in the middle.

`h-point: mean=   0.0000000000000000E+00 min=   0.0000000000000000E+00 max=   0.0000000000000000E+00 After set forcing fluxes%vprec`
`h-point: c= 0 After set forcing fluxes%vprec`
`h-point: mean=   1.2222794453150694E-20 min=  -2.6750820606277450E-17 max=   1.8644511331647810E-17 After set forcing fluxes%vprec`
`h-point: c=   1061651 After set forcing fluxes%vprec`

`h-point: c=  55749136 sw=  55816457 se=  55816457 nw=  55773806 ne=  55773806 Beginning of step_MOM  h`
`h-point: c=  55749136 sw=  55960457 se=  55960457 nw=  55917806 ne=  55917806 Beginning of step_MOM  h`

`h-point: c=  55749136 sw=  55816457 se=  55816457 nw=  55773806 ne=  55773806 Before steps  h`
`h-point: c=  55749136 sw=  55960457 se=  55960457 nw=  55917806 ne=  55917806 Before steps  h`

`h-point: c=  55749136 sw=  55816457 se=  55816457 nw=  55773806 ne=  55773806 Start set_viscous_BBL h`
`h-point: c=  55749136 sw=  55960457 se=  55960457 nw=  55917806 ne=  55917806 Start set_viscous_BBL h`

`h-point: c=  55749136 sw=  55816457 se=  55816457 nw=  55773806 ne=  55773806 Start predictor  h`
`h-point: c=  55749136 sw=  55960457 se=  55960457 nw=  55917806 ne=  55917806 Start predictor  h`

Any idea what is going wrong here?

**Comment by StephenGriffies** [comment]
Try removing water forcing and surface salinity restoring and see what happens.

**Comment by hmkhatri** [comment]
Thanks @StephenGriffies. I am running with uniform salinity. I am restoring salinity to 35 PSU at the surface and at the northern sponge. I will try switching it off and see if the problem goes away. Salinity restoration is unnecessary anyway.

**Comment by StephenGriffies** [comment]
Yep, as you say, if you are running with uniform salinity then no need to have any salinity boundary condition enabled.

**Comment by hmkhatri** [comment]
@Hallberg-NOAA @StephenGriffies Thanks for the comments. I ran a few test cases to pinpoint the source of the issue. I found out that the simulation started from a restart file only diverges if `RESTOREBUOY = True` is set. Otherwise, the solution does not change. I have verified that there is no issue with my input files which MOM reads for restoring surface buoyancy.

In order to check if this is an issue with my configuration or a general issue, I ran the rotating gravity current test (3D setup in Z*) case in MOM6-example with `RESTOREBUOY = True`. The problem appears in the test case too. I ran a fresh test simulation for 2 hrs in the first run. In another run, I restarted the simulation from T=1 hr and ran until T = 2 hrs. 

The output for the fresh run is,  

```
MOM Time       1.000     20: En 2.723625E-04, MaxCFL  0.00549, Mass 1.092186468882E+16, Salt  34.86939949537, Temp  -0.53952997875
    Total Energy: 4285A4D25D32B96A  2.9747060178791768E+12
    Total Mass:   1.0921864688817262E+16, Change:  -6.3040300875652971E+09 Error:  3.59181E-01 ( 3.3E-17)
    Total Salt:   3.8083886306875375E+14, Change:   1.9401851131439210E+02 Error:  1.94019E+02 ( 5.1E-13)
    Total Heat:  -2.3128743187036144E+19, Change:  -2.9843843953668020E+15 Error:  4.97786E+08 (-2.2E-11)
```
```
MOM Time       2.000     40: En 2.723697E-04, MaxCFL  0.00472, Mass 1.092185838479E+16, Salt  34.86941962183, Temp  -0.53959989551
    Total Energy: 4285A4F768DE2064  2.9747837081000488E+12
    Total Mass:   1.0921858384787170E+16, Change:  -6.3040300879244528E+09 Error:  2.54647E-05 ( 2.3E-21)
    Total Salt:   3.8083886306930712E+14, Change:   5.5343125437164304E+02 Error:  5.53431E+02 ( 1.5E-12)
    Total Heat:  -2.3131727049787236E+19, Change:  -2.9838627510922200E+15 Error:  1.80044E+09 (-7.8E-11)

```

The output for the run using restart file at T = 1 hr is,  

```
MOM Time       1.000      0: En 2.723625E-04, MaxCFL  0.00549, Mass 1.092186468882E+16, Salt  34.86939949537, Temp  -0.53952997875
    Total Energy: 4285A4D25D32B96A  2.9747060178791768E+12
    Total Mass:   1.0921864688817262E+16, Change:   0.0000000000000000E+00 Error:  0.00000E+00 ( 0.0E+00)
    Total Salt:   3.8083886306875375E+14, Change:   0.0000000000000000E+00 Error:  0.00000E+00 ( 0.0E+00)
    Total Heat:  -2.3128743187036144E+19, Change:   0.0000000000000000E+00 Error:  0.00000E+00 ( 0.0E+00)
```

```
MOM Time       2.000     20: En 2.723697E-04, MaxCFL  0.00472, Mass 1.092185839115E+16, Salt  34.86941960151, Temp  -0.53959982556
    Total Energy: 4285A4F768164485  2.9747837017045649E+12
    Total Mass:   1.0921858391149416E+16, Change:  -6.2976678389550467E+09 Error:  1.45025E-03 ( 1.3E-19)
    Total Salt:   3.8083886306930706E+14, Change:   5.5332864831542975E+02 Error:  5.53329E+02 ( 1.5E-12)
    Total Heat:  -2.3131724064583307E+19, Change:  -2.9808775471636100E+15 Error:  1.79972E+09 (-7.8E-11)
```

As it is clear from the output, answers change at T = 2.0 in the second run, which used a restart file.

I further looked at output in the debug mode to see where the solutions diverge. The first difference appears in `fluxes%vprec` calculations (below are the outputs for the fresh and restart runs, respectively, where the solutions diverge first), which I also mentioned in my previous comment.

```
h-point: mean=   1.5469334872094342E-06 min=   1.4136690481161624E-06 max=   1.7782038529591193E-06 After set forcing fluxes%vprec
h-point: c=   6636100 After set forcing fluxes%vprec
```

```
h-point: mean=   1.7678449097087771E-06 min=   1.5481942409400360E-06 max=   2.9824083650792626E-06 After set forcing fluxes%vprec
h-point: c=   6748171 After set forcing fluxes%vprec
```

I have tried my best but I am not able to resolve this issue myself. Does anyone have any idea on this? 
Thanks

**Comment by Hallberg-NOAA** [comment]
The way that you would debug this is to work your way back in the code to the point where fluxes%vprec is calculated in your case, and then add new checksum calls for all of the arrays that are used to calculate fluxes%vprec in your case, and then repeat the runs. You can also try using a debugger (like ddt) to look at the calculation of fluxes%vprec as it is being done.

  Is this case using the code in MOM_controlled_forcing.F90 to drive the model?

**Comment by hmkhatri** [comment]
Thanks for the comments, @Hallberg-NOAA . 

I am using the MOM6/dev-gfdl branch version. I have not done any changes to the source code. My impression is that the model is not using `MOM_controlled_forcing.F90`. Please correct me if I am mistaken.

**Comment by Hallberg-NOAA** [comment]
I did not expect it to be the case that you were using the control-theory forcing, but I wanted to raise this question in case it happened to be true.  That code has not been extensively used yet, so it would help to explain why you are finding something that has not been reported before. So we are back to the suggestion that you examine the fields that are used to set fluxes%vprec to figure out which of them is not being set correctly after the restart, and then how to fix it (presumably by adding  that field to the restart files).

**Comment by hmkhatri** [comment]
Setting `VARIABLE_BUOYFORCE = True` solves the issue.

**Comment by Hallberg-NOAA** [comment]
This issue has been addressed with PR#1192, which will issue a fatal error and an informative message if these inconsistent parameter settings are used.
